drv/psc-psu-update/src/main.rs

```diff
@@ -0,0 +1,262 @@
+        let now = Ticks::now();
+
+        self.last_checked = Some(now);
+        self.power_good = None;
+        self.firmware_matches = None;
+        self.firmware_revision = None;
+
+        if !dev.present() {
+            self.present = Some(false);
+
+            //
+            // If we are seeing our device as not present, we will clear our
+            // backoff value: if/when a PSU is plugged back in, we want to
+            // attempt to update it immediately if the firmware revision
+            // doesn't match our payload.
+            //
+            self.update_backoff = None;
+            return false;
+        }
+
+        self.present = Some(true);
+
+        match dev.power_good() {
+            Ok(power_good) => {
+                self.power_good = Some(power_good);
+
+                if !power_good {
+                    return false;
+                }
+            }
+            Err(err) => {
+                ringbuf_entry!(Trace::PowerGoodFailed(ndx, err));
+                return false;
+            }
+        }
+
+        match dev.firmware_revision() {
+            Ok(revision) => {
+                self.firmware_revision = Some(revision);
+
+                if revision == MWOCP68_FIRMWARE_REV {
+                    self.firmware_matches = Some(true);
+                    return false;
+                }
+
+                self.firmware_matches = Some(false);
+            }
+            Err(err) => {
+                ringbuf_entry!(Trace::FirmwareRevFailed(ndx, err));
+                return false;
+            }
+        }
+
+        if let (Some(started), Some(backoff)) =
+            (self.update_started, self.update_backoff)
+        {
+            if started + backoff > now {
+                ringbuf_entry!(Trace::BackingOff(ndx));
+                return false;
+            }
+        }
+
+        true
+    }
+
+    fn update_firmware(&mut self, dev: &Mwocp68, ndx: u8) {
+        ringbuf_entry!(Trace::AttemptingUpdate(ndx));
+        self.update_started = Some(Ticks::now());
+
+        self.update_backoff = match self.update_backoff {
+            Some(backoff) => Some(Ticks(backoff.0 * 2)),
+            None => Some(Ticks(60_000)),
+        };
+
+        let mut state = None;
+
+        loop {
+            match dev.update(state, MWOCP68_FIRMWARE_PAYLOAD) {
+                Err(err) => {
+                    //
+                    // We failed.  Record everything we can and leave.
+                    //
+                    ringbuf_entry!(Trace::UpdateFailed);
+                    ringbuf_entry!(Trace::UpdateFailedState(state));
+                    ringbuf_entry!(Trace::UpdateFailure(err));
+
+                    self.update_failure = Some((Ticks::now(), state, err));
+                    break;
+                }
+
+                Ok((UpdateState::UpdateSuccessful, _)) => {
+                    ringbuf_entry!(Trace::UpdateSucceeded);
+                    self.update_succeeded = Some(Ticks::now());
```

Comment:
I think we should reset `self.update_backoff` here, to catch the case of someone updating a PSU (possibly with failures, leading to backoff winding up), then swaps a new PSU into that slot in the 10-second window.